nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,848 @@
+        mat = xml.Element('Matrix')
+        if self.metadata:
+            mat.append(self.metadata._to_xml_element())
+        for mim in self.mims:
+            mat.append(mim._to_xml_element())
+        return mat
+
+
+class Cifti2Header(FileBasedHeader, xml.XmlSerializable):
+    ''' Class for Cifti2 header extension '''
+
+    # version = str
+
+    def __init__(self, matrix=None, version="2.0"):
+        FileBasedHeader.__init__(self)
+        xml.XmlSerializable.__init__(self)
+        if matrix is None:
+            self.matrix = Cifti2Matrix()
+        else:
+            self.matrix = matrix
+        self.version = version
+
+    def _to_xml_element(self):
+        cifti = xml.Element('CIFTI')
+        cifti.attrib['Version'] = str(self.version)
+        mat_xml = self.matrix._to_xml_element()
+        if mat_xml is not None:
+            cifti.append(mat_xml)
+        return cifti
+
+    @classmethod
+    def from_header(klass, header=None):
+        if header is None:
+            return klass()
+        if type(header) == klass:
+            return header.copy()
+        raise ValueError('header is not a Cifti2Header')
+
+    @classmethod
+    def may_contain_header(klass, binaryblock):
+        from .parse_cifti2_fast import _Cifti2AsNiftiHeader
+        return _Cifti2AsNiftiHeader.may_contain_header(binaryblock)
+
+
+class Cifti2Image(FileBasedImage):
+    # It is a Nifti2Image, but because Nifti2Image object
+    # contains both the *format* and the assumption that it's
+    # a spatial image, we can't inherit directly.
+    header_class = Cifti2Header
+    valid_exts = Nifti2Image.valid_exts
+    files_types = Nifti2Image.files_types
+    makeable = False
+    rw = True
+
+    def __init__(self, data=None, header=None, nifti_header=None):
+        self._header = header or Cifti2Header()
+        self.data = data
+        self.extra = nifti_header
+
+    def get_data(self):
+        return self.data
+
+    @classmethod
+    def from_file_map(klass, file_map):
+        """ Load a Cifti2 image from a file_map
+
+        Parameters
+        file_map : string
+
+        Returns
+        -------
+        img : Cifti2Image
+            Returns a Cifti2Image
+         """
+        from .parse_cifti2_fast import _Cifti2AsNiftiImage, Cifti2Extension
+        nifti_img = _Cifti2AsNiftiImage.from_file_map(file_map)
+
+        # Get cifti2 header
+        cifti_header = reduce(lambda accum, item:
+                              item.get_content()
+                              if isinstance(item, Cifti2Extension)
+                              else accum,
+                              nifti_img.get_header().extensions or [],
+                              None)
+        if cifti_header is None:
+            raise ValueError('Nifti2 header does not contain a CIFTI2 '
+                             'extension')
+
+        # Construct cifti image
+        cifti_img = Cifti2Image(data=np.squeeze(nifti_img.get_data()),
```

Comment:
The CIFTI format cares greatly about dimensions of length 1, and this is in fact common for dlabel and dscalar (on the first CIFTI dimension, which is the 5th NIfTI dimension) - this line is probably intended to deal with the four leading NIfTI dimensions that always have length 1 in a valid CIFTI file (this should be explicitly checked).  I advise against squeeze (as a general rule, even).